### PR TITLE
feat: support progressed output in `blob push` and `blob get`

### PR DESCRIPTION
**Pull request by qweeah**
**What this PR does / why we need it**:
This PR introduces terminal output for certain oras commands. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/oras-project/oras/issues/893

**Please check the following list**:
- [x]  Does the affected code have corresponding tests, e.g. unit test, E2E test?
- [ ]  Does this change require a documentation update?
- [ ]  Does this introduce breaking changes that would require an announcement or bumping the major version?
- [x]  Do all new files have an appropriate license header?

### CODE REVIEW

File: cmd/oras/internal/display/console/console.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+		return nil, err
+	}
+	return &Console{c}, nil
+}
+
+// Save saves the current cursor position.
+func (c *Console) Save() {
+	fmt.Fprint(c, aec.Hide)
+	// cannot use aec.Save since DEC has better compatilibity than SCO
+	fmt.Fprint(c, "\0337")
```
[the remark this answers is not in the document]
Comment (qweeah) [reply cmd/oras/internal/display/console/console.go]: A: should be more direct since `fmt.Fprint` also uses `c.Write`

File: cmd/oras/internal/display/progress/manager.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	return ch
+}
+
+// Close stops all status and waits for updating and rendering.
+func (m *manager) Close() error {
+	if m.closed() {
+		return errManagerStopped
+	}
+
+	// 1 wait for all model update done
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/progress/manager.go:138]: nit:
```suggestion
	// 1. wait for all model update done
```

File: cmd/oras/internal/display/track/reader.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	"os"
+	"sync/atomic"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras/cmd/oras/internal/display/progress"
+)
+
+type reader struct {
+	base         io.Reader
+	offset       atomic.Uint64
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/track/reader.go:29]: What's reason to make `offset` an atomic value?
> [!Note]
> All `io.Reader`s are assumed NOT go-routine safe.

File: cmd/oras/internal/option/common.go [hunk 4]
```diff
@@ -17,25 +17,44 @@ package option
...
 	"oras.land/oras/internal/trace"
 )
 
 // Common option struct.
 type Common struct {
 	Debug   bool
 	Verbose bool
+	TTY     *os.File
+
+	avoidTTY bool
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/common.go:35]: nit: consistency
```suggestion
	noTTY bool
```

File: cmd/oras/internal/option/common.go [hunk 5]
```diff
@@ -17,25 +17,44 @@ package option
...
 func (opts *Common) WithContext(ctx context.Context) (context.Context, logrus.FieldLogger) {
 	return trace.NewLogger(ctx, opts.Debug, opts.Verbose)
 }
+
+// Parse gets target options from user input.
+func (opts *Common) Parse() error {
+	f := os.Stderr
+	if !opts.avoidTTY && term.IsTerminal(int(f.Fd())) {
+		if opts.Debug {
+			return errors.New("cannot use --debug, add --noTTY to suppress terminal output")
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/common.go:55]:
```suggestion
			return errors.New("cannot use --debug, add --no-tty to suppress terminal output")
```

File: cmd/oras/internal/option/common.go [hunk 6]
```diff
@@ -17,25 +17,44 @@ package option
...
 }
 
 // WithContext returns a new FieldLogger and an associated Context derived from ctx.
 func (opts *Common) WithContext(ctx context.Context) (context.Context, logrus.FieldLogger) {
 	return trace.NewLogger(ctx, opts.Debug, opts.Verbose)
 }
+
+// Parse gets target options from user input.
+func (opts *Common) Parse() error {
+	f := os.Stderr
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/common.go:52]: Request documentation on setting `os.Stderr` as the output.

File: cmd/oras/internal/display/progress/manager.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+		for {
+			m.render()
+			select {
+			case <-m.renderDone:
+				m.render()
+				close(m.renderClosed)
+				return
+			case <-renderTicker.C:
+			}
+		}
```
[the remark this answers is not in the document]
Comment (qweeah) [reply cmd/oras/internal/display/progress/manager.go:81]: > Note: the above code is different from
Yes, the latter won't render output immediately after being started.

I think the code can be refactored to below so `manager.Close` only need to stop rendering and wait render to be closed.
```go
for {
	select {
	case <-m.renderDone:
		m.updating.Wait()
		m.render()
		close(m.renderClosed)
		return
	case <-renderTicker.C:
		m.render()
	}
}
```

File: cmd/oras/internal/option/common.go (hunk 6 above)
Comment (qweeah) [reply cmd/oras/internal/option/common.go:52]: What kind of docs are you suggesting? This is to accommodate the upcoming change for piped UX, so user can see the progress bar and use `-o json` at the same time.

File: cmd/oras/internal/display/track/reader.go (hunk 3 above)
Comment (qweeah) [reply cmd/oras/internal/display/track/reader.go:29]: Just trying to be go-routine safe here. Is it too much?

File: cmd/oras/internal/option/common.go (hunk 6 above)
Comment (qweeah) [reply cmd/oras/internal/option/common.go]: Added

File: cmd/oras/internal/display/track/reader.go (hunk 3 above)
Comment (qweeah) [reply cmd/oras/internal/display/track/reader.go:29]: Changed the implementation to non-atomic

File: cmd/oras/internal/display/progress/manager.go [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+func (m *manager) start() {
+	m.console.Save()
+	renderTicker := time.NewTicker(bufFlushDuration)
+	go func() {
+		defer m.console.Restore()
+		defer renderTicker.Stop()
+		for {
+			select {
+			case <-m.renderDone:
+				m.updating.Wait()
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/progress/manager.go:76]: `m.updating.Wait()` should be between line 138 and 139 before `close(m.renderDone)` so that the `manager` can keep rendering while waiting for all status to complete.

File: cmd/oras/internal/display/progress/status.go [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+		lenMargin = 0
+	}
+	return fmt.Sprintf("%s%s%s", left, strings.Repeat(" ", lenMargin), right), fmt.Sprintf("  └─ %s", s.descriptor.Digest.String())
+}
+
+// calculateSpeed calculates the speed of the progress and update last status.
+// caller must hold the lock.
+func (s *status) calculateSpeed() bytes {
+	now := time.Now()
+	secondsTaken := now.Sub(s.lastRenderTime).Seconds()
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/progress/status.go:152]: What will happen when `s.lastRenderTime` is zero? Should we use `startTime` instead?

File: cmd/oras/internal/display/track/reader.go [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+	if err != nil && err != io.EOF {
+		return n, err
+	}
+
+	r.offset = r.offset + int64(n)
+	if err == io.EOF {
+		if r.offset != r.descriptor.Size {
+			return n, io.ErrUnexpectedEOF
+		}
+		r.status <- progress.NewStatus(r.actionPrompt, r.descriptor, r.offset)
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/track/reader.go:95]: There is a perf issue regarding to the channel on line 95 and line 100.

Since `r.status` is **not buffered**, the read process is **blocked** if the manager fail to render in time.
Even if we make `r.status` is buffered, it is still possible that the buffer is full. Therefore, we need to make sure the `Read()` function will never be block by the `r.status <-` operation.

File: cmd/oras/internal/option/common.go [hunk 11]
```diff
@@ -17,25 +17,49 @@ package option
...
+	TTY     *os.File
+
+	noTTY bool
 }
 
 // ApplyFlags applies flags to a command flag set.
 func (opts *Common) ApplyFlags(fs *pflag.FlagSet) {
 	fs.BoolVarP(&opts.Debug, "debug", "d", false, "debug mode")
 	fs.BoolVarP(&opts.Verbose, "verbose", "v", false, "verbose output")
+	fs.BoolVarP(&opts.noTTY, "no-tty", "", false, "[Preview] avoid using stdout as a terminal")
```
[the remark this answers is not in the document]
Comment (shizhMSFT) [reply cmd/oras/internal/option/common.go:42]: /cc @FeynmanZhou seeking inputs on the flag name and description.

File: cmd/oras/internal/display/progress/humanize.go [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+	unit string
+}
+
+// ToBytes converts size in bytes to human readable format.
+func ToBytes(sizeInBytes int64) bytes {
+	f := float64(sizeInBytes)
+	if f < base {
+		return bytes{f, units[0]}
+	}
+	e := math.Floor(math.Log(f) / math.Log(base))
```
[the remark this answers is not in the document]
Comment (qweeah) [reply cmd/oras/internal/display/progress/humanize.go]: Let's set it to `len(units)-1`

File: cmd/oras/internal/display/progress/manager.go (hunk 8 above)
Comment (qweeah) [reply cmd/oras/internal/display/progress/manager.go:76]: Nice catch. Will revert.

File: cmd/oras/internal/display/progress/status.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (qweeah) [reply cmd/oras/internal/display/progress/status.go:153]: How about setting it to a value no smaller than `bufFlushDuration`?

File: cmd/oras/internal/display/track/reader.go (hunk 10 above)
Comment (qweeah) [reply cmd/oras/internal/display/track/reader.go:95]: Wow, this trick is amazing!!! In this case we only need the buffer size to be 2? Anyway old status doesn't need to be updated to the manager?

File: cmd/oras/internal/display/progress/manager.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (qweeah) [reply cmd/oras/internal/display/progress/manager.go]: Confirmed that using `m.updating.Wait()` during rendering will make UX update stuck in some edge cases.

File: cmd/oras/internal/option/common.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (FeynmanZhou) [reply cmd/oras/internal/option/common.go]: If the `no-tty` only applies to the progress bar feature in ORAS so far, how about rewording the description to be much simpler and straightforward? 

```
fs.BoolVarP(&opts.noTTY, "no-tty", "", false, "[Preview] do not show progress output")
````

File: cmd/oras/root/blob/push.go [hunk 13]
```diff
@@ -145,3 +140,32 @@ func pushBlob(ctx context.Context, opts pushBlobOptions) (err error) {
...
+		defer trackedReader.StopManager()
+		trackedReader.Start()
+		r = trackedReader
+		if err := t.Push(ctx, desc, r); err != nil {
+			return err
+		}
+		trackedReader.Done()
+	}
+	return nil
+}
```
Comment (shizhMSFT) [inline cmd/oras/root/blob/push.go:171]: nit:
```suggestion
func (opts *pushBlobOptions) doPush(ctx context.Context, t oras.Target, desc ocispec.Descriptor, r io.Reader) error {
	if opts.TTY == nil {
		// none tty output
		if err := display.PrintStatus(desc, "Uploading", opts.Verbose); err != nil {
			return err
		}
		if err := t.Push(ctx, desc, r); err != nil {
			return err
		}
		return display.PrintStatus(desc, "Uploaded ", opts.Verbose)
	} 
	
	// tty output
	trackedReader, err := track.NewReader(r, desc, "Uploading", "Uploaded ", opts.TTY)
	if err != nil {
		return err
	}
	defer trackedReader.StopManager()
	trackedReader.Start()
	r = trackedReader
	if err := t.Push(ctx, desc, r); err != nil {
		return err
	}
	trackedReader.Done()
	return nil
}
```

File: cmd/oras/internal/display/progress/status.go [hunk 14]
```diff
@@ -147,15 +148,16 @@ func (s *status) String(width int) (string, string) {
 
 // calculateSpeed calculates the speed of the progress and update last status.
 // caller must hold the lock.
-func (s *status) calculateSpeed() bytes {
+func (s *status) calculateSpeed() humanize.Bytes {
 	now := time.Now()
 	secondsTaken := now.Sub(s.lastRenderTime).Seconds()
+	secondsTaken = math.Max(secondsTaken, float64(bufFlushDuration.Milliseconds())/1000)
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/progress/status.go:154]: Instead of using `math.Max`, can you use `if secondsTaken == 0` instead as it is an extremely rare case?

File: cmd/oras/root/blob/fetch.go [hunk 15]
```diff
@@ -147,40 +147,35 @@ func (opts *fetchBlobOptions) doFetch(ctx context.Context, src oras.ReadOnlyTarg
...
-		// save blob content into the local file if the output path is provided
-		file, err := os.Create(opts.outputPath)
-		if err != nil {
-			return ocispec.Descriptor{}, err
+			writer = file
+			defer func() {
+				if err := file.Close(); fetchErr == nil {
+					fetchErr = err
+				}
+			}()
```
Comment (shizhMSFT) [inline cmd/oras/root/blob/fetch.go:162]: nit: coding style: `defer` first after error handling then other code
```suggestion
			defer func() {
				if err := file.Close(); fetchErr == nil {
					fetchErr = err
				}
			}()
			writer = file
```

File: cmd/oras/internal/display/progress/status.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (shizhMSFT) [reply cmd/oras/internal/display/progress/status.go:153]: See https://github.com/oras-project/oras/pull/1113#discussion_r1357885268

Comment (shizhMSFT) [reply cmd/oras/internal/display/progress/status.go:153]: @qweeah This still need to be addressed.

**Review by shizhMSFT — APPROVED** [review]
LGTM
